src/notify_telegram.c

```diff
@@ -0,0 +1,871 @@
+    char url[MAX_URL_SIZE] = "";
+    if (proxy_url) {
+      snprintf(url, MAX_URL_SIZE, "%s%s/deleteWebhook", proxy_url, bot_token);
+    } else {
+      snprintf(url, MAX_URL_SIZE,
+               "https://api.telegram.org/bot%s/deleteWebhook", bot_token);
+    }
+    char params[MAX_PARAMS_SIZE] = "";
+    CURL *handle = curl_easy_init();
+    curl_easy_setopt(handle, CURLOPT_POSTFIELDS, params);
+    curl_easy_setopt(handle, CURLOPT_URL, url);
+    curl_easy_setopt(handle, CURLOPT_WRITEFUNCTION, tg_curl_write_callback);
+    curl_easy_setopt(handle, CURLOPT_WRITEDATA, (void *)&buf);
+    pthread_mutex_lock(&telegram_lock);
+    CURLcode response_code = curl_easy_perform(handle);
+    pthread_mutex_unlock(&telegram_lock);
+    curl_easy_cleanup(handle);
+    DEBUG("notify_telegram: curl response = %s", buf.response);
+    if (response_code != CURLE_OK) {
+      ERROR("notify_telegram: curl_easy_perform failed.");
+      sfree(buf.response);
+      return -1;
+    }
+  }
+
+  return notify_telegram_read();
+}
+
+static int notify_telegram_shutdown(void) {
+  curl_global_cleanup();
+
+  if (httpd) {
+    MHD_stop_daemon(httpd);
+    httpd = NULL;
+  }
+
+  return 0;
+}
+
+static int notify_telegram_config(const char *key, const char *value) {
+  if (strcasecmp(key, "RecipientChatID") == 0) {
+    char **tmp;
+    tmp = realloc(recipients, (recipients_len + 1) * sizeof(char *));
+    if (tmp == NULL) {
+      ERROR("notify_telegram: realloc failed.");
+      return -1;
+    }
+    recipients = tmp;
+    recipients[recipients_len] = strdup(value);
+    if (recipients[recipients_len] == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+    recipients_len++;
+  } else if (strcasecmp(key, "BotToken") == 0) {
+    sfree(bot_token);
+    bot_token = strdup(value);
+    if (bot_token == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "ProxyURL") == 0) {
+    sfree(proxy_url);
+    proxy_url = strdup(value);
+    if (proxy_url == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "DisableGettingUpdates") == 0) {
+    disable_getting_updates = IS_TRUE(value);
+  } else if (strcasecmp(key, "WebhookURL") == 0) {
+    sfree(webhook_url);
+    webhook_url = strdup(value);
+    if (webhook_url == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "WebhookHost") == 0) {
+    sfree(webhook_host);
+    webhook_host = strdup(value);
+    if (webhook_host == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "WebhookPort") == 0) {
+    sfree(webhook_port);
+    webhook_port = strdup(value);
+    if (webhook_port == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "MHDDaemonHost") == 0) {
+#if MHD_VERSION >= 0x00090000
```

Comment:
Checks for earlier MHD version keys could be done separately from the strdup() loop.  Or the looped table could have member indicating for which MHD version give key is.